### ISSUE TITLE
problems signing in 12501 when doing ios + android dev

### ISSUE THREAD

**Issue by cpup22** [EddyVerbruggen/cordova-plugin-googleplus#327]
I noticed I can get the plugin working great on android and then I test/run/build my cordova app for ios (where it works fine on ios) then when I go to run/build on android I get the 12501 error.  Only way to solve is to uninstall the plugin and reinstall.  Which can be frustrating when developing for both platforms constantly having to uninstall/reinstall the plugin.

**Comment by limaneto** [comment]
have you fixed that ?

**Comment by SilvaCoder** [comment]
@JohnMcAvinue Exactly the same here. I tried all possible combinations in mind.

**Comment by ilkkanisula** [comment]
@JohnMcAvinue I'm also getting 12501 (SIGN_IN_CANCELLED) error. Maybe it doesn't work with the firebase authentication?

**Comment by JohnMcAvinue** [comment]
@ilkkanisula That could well be it. You would think the developer would indicate if this is the case though. Have you tried it with the old way?

**Comment by ilkkanisula** [comment]
This plugin does about the same but uses some firebase sdk methods also:
https://github.com/blakgeek/cordova-plugin-firebase-auth-google/blob/master/src/android/com/blakgeek/cordova/plugin/FirebaseAuthPlugin.java

**Comment by JohnMcAvinue** [comment]
I created new clients for Android and Web Application in Google Credentials Manager (not Firebase) and tried with those but still getting a 12501. This is almost certainly an issue with the plugin or documentation as I cannot get it to work

**Comment by ilkkanisula** [comment]
Anyone got this working yet?

**Comment by SilvaCoder** [comment]
Nope, i had to use the "web" way of doing that.

If anyone wants to jump over this issue and implement he "web" way of getting this done, for futher work on the real solution just hit me and i'll help, or post a demo working properly.

**Comment by SilvaCoder** [comment]
But i want to say that im starting to think this native G+ SignIn will work only after signing the app and sending it to production/publishment.

**Comment by JohnMcAvinue** [comment]
I've luckily been able to park this feature for now, and am hoping the developer can come back to help us with this until I actually need it. No point in doing the Web way cos Google are depreciating it in a few months

**Comment by SilvaCoder** [comment]
@JohnMcAvinue What? Google cant deprecate the web G+ Sign In. What i heard, and indeed the Google did, was not allowing a webview handle signin procress anymore. But i could get around this easily. It does not make sense remove the web signin cause sites will still have this feature on.

**Comment by JohnMcAvinue** [comment]
@SilvaCoder Yes I thought you were talking about OAuth using InAppBrowser which is being depreciated

**Comment by SilvaCoder** [comment]
But im using InAppBrowser with Crosswalk without webview in order to make it work.
But yes, the normal procress with InAppBrowser is deprecated, but like i said i could get around this and its working pretty well, cause the only thing that Google did was upload a file, and actually webviews cant handle that, thats how they broke the old normal procress. For now is better than nothing, but im still working in the real solution.

**Comment by ilkkanisula** [comment]
I have also tried to sign in the app with production key and it's still not working properly.

**Comment by SilvaCoder** [comment]
@ikkanisula Thanks for response ! I'll not jump directly to it then.

**Comment by ilkkanisula** [comment]
So now it's working fine. You need to sign your app with release key using commands:
```
jarsigner -verbose -sigalg SHA1withRSA -digestalg SHA1 -storepass PASSWORD -keystore your.keystore platforms/android/build/outputs/apk/android-release-unsigned.apk release

zipalign -v 4 platforms/android/build/outputs/apk/android-release-unsigned.apk android-release-signed.apk
```

**Comment by JohnMcAvinue** [comment]
I'm using ionic cloud to build and have used my release keystore but it doesn't work.

You would imagine using that would mean it's signed?

John

**Comment by sammuggPS** [comment]
Make sure your release SHA1 is the same as what you've used in the Developers Console.

**Comment by JohnMcAvinue** [comment]
I'm 100% certain it is and the package name matches. See what I've tried a few posts up.

John

**Comment by Substractive** [comment]
I managed to get it running did everything like documentation said it was bit tricky in developer console I created oauth web and android version. For android version I followed steps for SHA and package. After creating those two I installed plugin with reversed id of android version of app created in developer console. After installing the app in my .js file in webClientId I put ID of the web app created in google developer console and now it all works. 
Hopefully this helps.

**Comment by JohnMcAvinue** [comment]
Which keystore did you use for the SHA1?

**Comment by Substractive** [comment]
Android debug keystore found in .android folder (on my mac)

**Comment by SilvaCoder** [comment]
So, just for settle down things about this issue. If you plan to make it work on debug you must sign the app with a labeled debug key and this key must be in firebase app on the place where you can put your cert fingerprint. And in order to make it work in production you must sign the app with a labeled release key and put it on developers console when publishing the app. Is this correct ?

**Comment by gylippus** [comment]
Just wanted to add here that I was facing the same 12501 errors. I'm using Firebase and it appears that the generated "Web SDK configuration" code for Google Sign-In authorization there did not actually match up with what was in the onsole.developers.google.com OAuth 2.0 client IDs for Web Client. I took the Web Client ID from Google Credentials Manager and replaced the one in Firebase. Everything worked right away after that passing the same one to the plugin in my js.

**Comment by JTronLabs** [comment]
You can automatically sign a Cordova app when running on your local machine with 

`cordova run android --release`

and then adding a [build.json](https://cordova.apache.org/docs/en/latest/guide/platforms/android/#using-buildjson) file next to your config. xml. If you want to commit this file you can leave out your keystore's passwords and popup will ask you to enter keystore and alias passwords.

**Comment by Barathwaja** [comment]
i was able to get rid of error 12501.. please give the SHA-1 key correctly ...

1 : create an ionic proj. 
2 : after that add the plugin and stuffs given by Eddy
3 : then create android platform and add keystore (if release build) or debug.keystore available in this path %USERPROFILE%/.android/debug.keystore.
4 : google how to find SHA-1 in in debug mode and copy and paste in reqd. proj
5 : finally build apk and it wll work,..

**Comment by Ross-Rawlins** [comment]
I tried @Barathwaja steps and I still get 12501. I created a debug and a release keystore and then I created two separate projects in google console and neither works. I have re tried the process from start to finish 3 times and its still failing.

**Comment by sammuggPS** [comment]
Double check that your project is actually building with the keystore you intend it to. Sometimes, especially when using Ionic and Phonegap, they have a separate debug keystore that overrides if you don't specify exactly which one to use.

**Comment by Barathwaja** [comment]
Please be sure of SHA-1 coz if that's mismatched it will not work.. Am sure of that @Ross-Rawlins

**Comment by Ross-Rawlins** [comment]
Hey Guys someone in the cordova slack group told me to use the android client id and not the web client id once I swapped them then it all worked 100%;

**Comment by reno77** [comment]
Ok.. finally got it working after 2 hours.
You need to use the reversed Android client key in the config.xml 
and the Web client (auto created by Google Service) key in the window.plugins.googleplus.login call ..

I've created 2 Android client keys in the Google API manager, one for Debug and one for Release, and change the reversed client key in config.xml depending on which version I'm compiling currently.

**Comment by J-Yen** [comment]
I fixed the problem by specifying the location of the keystore:
https://www.webniraj.com/2016/11/21/cordova-changing-the-default-debug-keystore-for-android-applications/

**Comment by MoePad** [comment]
Just to let you (and all following guys) know:   I had the same problem because I didn't read the documentation of Eddy well enough. **The SCOPES are SPACE SEPERATED**, not comma seperated. This also results in 12501.

**Comment by Daedgomez** [comment]
@reno77 solution works for me.
